### PR TITLE
Feat free properties as filter zweite welle

### PR DESCRIPTION
**Pull request by laurgarn**

### CODE REVIEW

**Comment by dbraquart** [comment]
"zweite welle" in PR title ??

File: src/utils/equipment-types.js [hunk 1]
```diff
@@ -87,6 +75,19 @@ const FreeProps = {
         type: filteredTypes.freeProperties,
     },
 };
+const FreePropsS = {
```
Comment (dbraquart) [inline src/utils/equipment-types.js:78]: S for substation ? you can be more explicit in naming

File: src/components/dialogs/create-filter-dialog.js [hunk 2]
```diff
@@ -112,14 +112,16 @@ const CreateFilterDialog = ({
...
+    const [validationsCount, setValidationsCount] = useState(0);
     const [tableValues, setTableValues] = useState([]);
     const [isGeneratorOrLoad, setIsGeneratorOrLoad] = useState(false);
     const [isFilterCreation, setIsFilterCreation] = useState(false);
     const [equipmentType, setEquipmentType] = useState(null);
     const [name, setName] = useState('');
     const [id, setId] = useState('');
-    const handleCallback = (criteriaFilter) => {
-        if (criteriaFilter) {
+    const handleCallback = (criteriaFilter, veto) => {
```
Comment (dbraquart) [inline src/components/dialogs/create-filter-dialog.js:122]: what this flag "veto" really means ?

File: src/components/dialogs/create-filter-dialog.js [hunk 3]
```diff
@@ -209,8 +211,9 @@ const CreateFilterDialog = ({
     }, [tableValues]);
 
     const handleValidation = () => {
-        //To manage the case when we never tried to enter a name
+        setValidationsCount((prev) => prev + 1);
```
Comment (dbraquart) [inline src/components/dialogs/create-filter-dialog.js:214]: I dont understand the purpose of this counter ? maybe we can discuss about it, dont reply here.

File: src/components/dialogs/criteria-based-filter-dialog-content.js [hunk 4]
```diff
@@ -83,12 +84,80 @@ export const FilterTypeSelection = ({
     );
 };
 
+const backToFrontTweak = (response) => {
```
Comment (dbraquart) [inline src/components/dialogs/criteria-based-filter-dialog-content.js:91]: maybe you could comment these 2 functions. They are very "technical", it's hard to follow what they really do

File: src/components/dialogs/criteria-based-filter-dialog-content.js [hunk 5]
```diff
@@ -115,14 +184,12 @@ export const CriteriaBasedFilterDialogContent = ({
             if (contentType === ElementType.FILTER) {
                 getFilterById(id)
                     .then((response) => {
-                        setInitialFilter(response);
-                        setEquipmentType(
-                            response.equipmentFilterForm.equipmentType
-                        );
+                        setInitialFilter(backToFrontTweak(response));
+                        let eType = response.equipmentFilterForm.equipmentType;
```
Comment (dbraquart) [inline src/components/dialogs/criteria-based-filter-dialog-content.js:200]: etype could be const

File: src/components/dialogs/expandable-criteria.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+export const useExpandableCriteria = ({
+    id,
+    labelAddValue,
+    Field,
+    fieldProps,
+    initialValues,
+    onChange,
+    validateItems,
+}) => {
+    // const classes = useStyles();
```
Comment (dbraquart) [inline src/components/dialogs/expandable-criteria.js]: remove it

File: src/components/dialogs/expandable-criteria.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                        </Button>
+                    </span>
+                </Grid>
+            </Grid>
+        );
+    }, [
+        values,
+        fieldProps,
+        // classes.button,
+        // classes.deleteButton,
```
Comment (dbraquart) [inline src/components/dialogs/expandable-criteria.js]: remove comments

File: src/components/dialogs/free-properties-based-filter.js [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+    }, [initialValue]);
+    const validationsCountRef = useRef();
+
+    const onPropertiesArrayChange = useCallback(
+        (arr) => {
+            const obj = !arr
+                ? {}
+                : Object.fromEntries(arr.map((p) => [p.name || '', p.values]));
+            const vetoIdx = arr.findIndex((p) => !p.name || !p?.values?.length);
+            onChange(obj, vetoIdx >= 0);
```
Comment (dbraquart) [inline src/components/dialogs/free-properties-based-filter.js]: here is veto again ... a comment could help. What are we doing in this function ?

**Review by dbraquart — CHANGES_REQUESTED** [review]
review: a few comments
tests: 2 ko
a) unable to add/create a filtering by property (+ button), for eqpt "HVDC Lines".
The sent Json sounds bad, it does not use freeProperties1 / freeProperties2 like with the Lines.

b)
1- create Filter on line (or hvdc line), with at least one properties line :region/east/west + validate
2- edit it, switch to another eqpt => crash
 CardErrorBoundary caught:  TypeError: value is undefined
    useAutocomplete useAutocomplete.js:916
    Autocomplete Autocomplete.js:423

remark: when I enter twice the same properties name, I still can validate the Form (the second properties line is deleted)

Console warning: I got 3 warnings:
1) Warning: Failed prop type: Invalid prop `columns[0].minWidth` of type `string` supplied to `MuiVirtualizedTable`, expected `number`.
2)  Warning: Failed prop type: Invalid prop `minWidth` of type `string` supplied to `Column`, expected `number`.
3)  Warning: Cannot update a component (`CriteriaBasedFilterDialog`) while rendering a different component (`SingleFilter`).
 To locate the bad setState() call inside `SingleFilter`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render

File: src/utils/equipment-types.js (hunk 1 above)
Comment (laurgarn) [reply src/utils/equipment-types.js:78]: We are happy with '1' and '2'. Why could we not be happy with just this 'S' that has just one local usage ?

File: src/components/dialogs/create-filter-dialog.js (hunk 2 above)
Comment (laurgarn) [reply src/components/dialogs/create-filter-dialog.js:122]: "Veto" is usually used to tell "I/we don't want", see https://en.wiktionary.org/wiki/veto. And this is what it means here, as it is use also in https://docs.oracle.com/javase/7/docs/api/index.html?java/beans/VetoableChangeListener.html
The method in itself is not that long and it can quite easily be infered that if veto is true, no filter is to be saved (and even be reset).

File: src/components/dialogs/criteria-based-filter-dialog-content.js (hunk 5 above)
Comment (laurgarn) [reply src/components/dialogs/criteria-based-filter-dialog-content.js:200]: Yes, it could. But it would trigger the 80 characters per line limit that bother me more.

File: src/components/dialogs/create-filter-dialog.js (hunk 3 above)
Comment (laurgarn) [reply src/components/dialogs/create-filter-dialog.js:214]: It carries the information that validation number has changed or not, allowing to not mark as a problem something like an empty property line after it has been added but has it marked after a new validation.

File: src/components/dialogs/criteria-based-filter-dialog-content.js (hunk 4 above)
Comment (laurgarn) [reply src/components/dialogs/criteria-based-filter-dialog-content.js:91]: It just adapts the data structure from what is received from the backend to the form that the component can handle.

File: src/components/dialogs/expandable-criteria.js (hunk 6 above)
Comment (laurgarn) [reply src/components/dialogs/expandable-criteria.js]: Well spotted

File: src/components/dialogs/expandable-criteria.js (hunk 7 above)
Comment (laurgarn) [reply src/components/dialogs/expandable-criteria.js]: Thx

**Comment by laurgarn** [comment]
> review: a few comments tests: 2 ko
> a) unable to add/create a filtering by property (+ button), for eqpt "HVDC Lines". The sent Json sounds bad, it does not use freeProperties1 / freeProperties2 like with the Lines.

addressed and corrected.

> b) 1- create Filter on line (or hvdc line), with at least one properties line :region/east/west + validate
>     2- edit it, switch to another eqpt => crash CardErrorBoundary caught: TypeError: value is undefined useAutocomplete useAutocomplete.js:916 Autocomplete Autocomplete.js:423

I'll address this. To trigger this, we need to switch to a none line or hvdc

> remark: when I enter twice the same properties name, I still can validate the Form (the second properties line is deleted)
> 
> Console warning: I got 3 warnings:
> 
>     1. Warning: Failed prop type: Invalid prop `columns[0].minWidth` of type `string` supplied to `MuiVirtualizedTable`, expected `number`.
> 
>     2. Warning: Failed prop type: Invalid prop `minWidth` of type `string` supplied to `Column`, expected `number`.

These first two have nothing to do with this PR.
They have with me, because I told Walid to use a css compatible size as column minWidth, forgetting our MUIVirtualizedTable sizes method does not handle that smartly enough.
But this has nothing to do we the changed code of this PR.

>     3. Warning: Cannot update a component (`CriteriaBasedFilterDialog`) while rendering a different component (`SingleFilter`).
>        To locate the bad setState() call inside `SingleFilter`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render

File: src/utils/equipment-types.js (hunk 1 above)
Comment (dbraquart) [reply src/utils/equipment-types.js:78]: hum.. not so local, you have a corresponding component named FreePropertiesS

File: src/components/dialogs/free-properties-based-filter.js [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+                            />
+                        ))
+                    }
+                />
+            </FormControl>
+        </>
+    );
+};
+
+const makeFreePropertiesProblemsMap = (values, full, prevErrors) => {
```
Comment (dbraquart) [inline src/components/dialogs/free-properties-based-filter.js]: this function is a clone of makeFreeProperties2ProblemsMap, except one line.
could we use a single function ? (proposal)

**Review by dbraquart — CHANGES_REQUESTED** [review]

File: src/components/dialogs/free-properties-based-filter.js (hunk 9 above)
Comment (dbraquart) [reply src/components/dialogs/free-properties-based-filter.js]: review: ok (1 remark below)
console: ok
tests : ok

**Review by dbraquart — APPROVED** [review]
review: ok
console: ok
tests : ok

**Review by ayolab — APPROVED** [review]
